### ISSUE TITLE
gevent could use some monitoring/debugging tools

### ISSUE THREAD

**Issue by jamadden** [gevent/gevent#1021]
Probably inspired by https://github.com/mozilla-services/mozservices/blob/master/mozsvc/gunicorn_worker.py but generic for any gevent-based application.

**Comment by jamadden** [comment]
Here's a debugging function I find useful sometimes:
```python
def dump_stacks():
    """
    Request information about the running threads of the current process.

    :return: A sequence of text lines detailing the stacks of running
            threads and greenlets. (One greenlet will duplicate one thread,
            the current thread and greenlet.)
    """
    dump = []

    import gc   
    # threads
    import threading  # Late import this stuff because it may get monkey-patched
    import sys
    import traceback
    try:
        from greenlet import greenlet

    except ImportError:
        greenlet = None

    threads = {th.ident: th.name for th in threading.enumerate()}

    for thread, frame in sys._current_frames().items():
        dump.append('Thread 0x%x (%s)\n' % (thread, threads.get(thread)))
        dump.append(''.join(traceback.format_stack(frame)))
        dump.append('\n')

    # greenlets

    # if greenlet is present, let's dump each greenlet stack
    # Use the gc module to inspect all objects to find the greenlets
    # since there isn't a global registry
    for ob in gc.get_objects():
        if not isinstance(ob, greenlet):
            continue
        if not ob:
            continue  # not running anymore or not started
        dump.append('Greenlet %s\n' % ob)
        dump.append(''.join(traceback.format_stack(ob.gr_frame)))
        dump.append('\n')

    return dump
```

**Comment by jamadden** [comment]
cf #755

**Comment by jamadden** [comment]
cf #1123 `dump_stacks` is now `gevent.util.format_run_info` (I wish I had a better name) and prints a reasonably nice-looking tree of all greenlet objects, their locals and spawn tree locals, their stacks and spawn stacks.

**Comment by arcivanov** [comment]
> I wish I had a better name
`dump_traces()` ?

**Comment by jamadden** [comment]
Thanks! That’s pretty similar to the original name. The issue I had with that was that it doesn’t convey that this now includes much more than just stack traces.

**Comment by kurtbrose** [comment]
"greenlet friendly" console: https://github.com/paypal/support/blob/master/support/async.py#L470-L483

```
class GreenConsole(code.InteractiveConsole):
    @io_bound
    def raw_input(self, prompt=""):
        return code.InteractiveConsole.raw_input(self, prompt)


def start_repl(local=None, banner="infra REPL (exit with Ctrl+C)"):
    gevent.spawn(GreenConsole().interact, banner)


def greenify(banner="REPL is now greenlet friendly (exit with Ctrl+C)"):
    import __main__
    GreenConsole(__main__.__dict__).interact(banner)
```

this worked very well with a server -- so in addition to starting to serve traffic you also start a REPL on stdin, makes it easy to dig into objects while the thing is running

**Comment by kurtbrose** [comment]
not really gevent-specific, but kind of a "socket garbage disposal" function -- try to shutdown cleanly if possible but under no circumstances raise an error just get rid of the socket

https://github.com/paypal/support/blob/master/support/async.py#L422-L449

```
def killsock(sock):
    """Attempts to cleanly shutdown a socket. Regardless of cleanliness,
    ensures that upon return, the socket is fully closed, catching any
    exceptions along the way. A safe and prompt way to dispose of the
    socket, freeing system resources.
    """
    if hasattr(sock, '_sock'):
        ml.ld("Killing socket {0}/FD {1}", id(sock), sock._sock.fileno())
    else:
        ml.ld("Killing socket {0}", id(sock))
    try:
        # TODO: better ideas for how to get SHUT_RDWR constant?
        sock.shutdown(gevent.socket.SHUT_RDWR)
    except gevent.socket.error:
        pass  # just being nice to the server, don't care if it fails
    except Exception as e:
        log_rec = context.get_context().log.info("SOCKET", "SHUTDOWN")
        log_rec.failure('error ({exc}) shutting down socket: {socket}',
                        socket=sock, exc=e)
    try:
        sock.close()
    except gevent.socket.error:
        pass  # just being nice to the server, don't care if it fails
    except Exception as e:
        log_rec = context.get_context().log.info("SOCKET", "CLOSE")
        log_rec.failure('error ({exc}) closing socket: {socket}',
                        socket=sock, exc=e)
```

**Comment by kurtbrose** [comment]
`cpu_bound` decorator to dispatch to (single) thread for cpu_bound things

the difference between a "cpu" task and a threadpool task is you only ever want one cpu thread so multiple cpu tasks don't just step on each other

https://github.com/paypal/support/blob/master/support/async.py#L356-L381

the next step from that is `cpu_bound_if` decorator

https://github.com/paypal/support/blob/master/support/async.py#L384-L410

this kicks in, lets say you are encrypting or BASE-64-ing some data; it is not uncommon for message sizes to vary hugely, maybe for a 100K message you definitely want to kick it over to a thread, but for a 100-byte message you don't want to pay the ~2ms overhead of the thread dispatch

**Comment by kurtbrose** [comment]
also a more general threadpooldispatcher https://github.com/paypal/support/blob/master/support/async.py#L172-L216

both the threadpool and cpu_thread try to avoid child thread dispatching; so cpu_bound calls cpu_bound or threadpool calls threadpool, they will detect that they are already running "in the background" and not do another layer of Q / dispatching

**Comment by kurtbrose** [comment]
https://github.com/paypal/support/blob/master/support/async.py#L96-L152

staggered retries -- instead of timing out, kick off a second and third request that run in parallel (a different kind of spawn function -- you hand it the function to spawn it spawns it multiple times if it doesn't finish fast enough)

this one is a little subtle -- you are trying to cut off the long tail of server response times from the client side without putting too much extra load on the server

say the 99% response time is 10x the average; you set the retry time at 99%, so that causes 1% extra load in return for which your 99% time goes from 10x to 2x (because the second call is another roll of the dice) -- the third retry you add 0.01% and cut it down to 3x

there are risks, but in general it is a great technique for managing long tail

see also "happy eyeballs" for trying multiple addresses in parallel:
https://tools.ietf.org/html/rfc8305

**Comment by kurtbrose** [comment]
oops sorry drifting off topic of debugging tools :-)

**Comment by mahmoud** [comment]
@kurtbrose https://www.youtube.com/watch?v=jmo0M2_OWsU :rofl:

**Comment by jamadden** [comment]
Thanks everyone for the comments! I'm going to consider this issue closed as far as 1.3 goes. I've [referenced it in the docs](http://www.gevent.org/monitoring.html) to give the not-quite-debugging, not-quite-gevent-specific some extra visibility.
